command.go

```diff
@@ -0,0 +1,234 @@
+// Args should not include the name of the program: os.Args[1:], not os.Args.
+func (cmd *Command) Parse(args []string, options ...Option) error {
+	// Initial validation and safety checks.
+	if cmd.Name == "" {
+		return fmt.Errorf("name is required")
+	}
+	if cmd.isParsed {
+		return fmt.Errorf("%s: %w", cmd.Name, ErrAlreadyParsed)
+	}
+
+	// If no flag set was given, set an empty default, so -h, --help works.
+	if cmd.Flags == nil {
+		cmd.Flags = NewFlags(cmd.Name)
+	}
+
+	// Parse this command's flag set from the provided args.
+	if err := parseFlags(cmd.Flags, args, options...); err != nil {
+		cmd.selected = cmd // allow GetSelected to work even with errors
+		return fmt.Errorf("%s: %w", cmd.Name, err)
+	}
+
+	// If the parse was successful, mark the command as parsed.
+	cmd.isParsed = true
+
+	// Set this command's args to the args left over after parsing.
+	cmd.args = cmd.Flags.GetArgs()
+
+	// If there were any args, we might need to descend to a subcommand.
+	if len(cmd.args) > 0 {
+		first := cmd.args[0]
+		for _, subcommand := range cmd.Subcommands {
+			if strings.EqualFold(first, subcommand.Name) {
+				cmd.selected = subcommand
+				subcommand.parent = cmd
+				return subcommand.Parse(cmd.args[1:], options...)
+			}
+		}
+	}
+
+	// We didn't find a matching subcommand, so we selected ourselves.
+	cmd.selected = cmd
+
+	// Parse complete.
+	return nil
+}
+
+// Run the exec function of the command selected during the parse phase, passing
+// the args left over after parsing. Calling run without first calling parse
+// will result in an error.
+func (cmd *Command) Run(ctx context.Context) error {
+	switch {
+	case !cmd.isParsed:
+		return ErrNotParsed
+	case cmd.isParsed && cmd.selected == nil:
+		return ErrNotParsed
+	case cmd.isParsed && cmd.selected == cmd && cmd.Exec == nil:
+		return fmt.Errorf("%s: %w", cmd.Name, ErrNoExec)
+	case cmd.isParsed && cmd.selected == cmd && cmd.Exec != nil:
+		return cmd.Exec(ctx, cmd.args)
+	default:
+		return cmd.selected.Run(ctx)
+	}
+}
+
+// ParseAndRun calls parse and then, on success, run.
+func (cmd *Command) ParseAndRun(ctx context.Context, args []string, options ...Option) error {
+	if err := cmd.Parse(args, options...); err != nil {
+		return fmt.Errorf("parse: %w", err)
+	}
+
+	if err := cmd.Run(ctx); err != nil {
+		return fmt.Errorf("run: %w", err)
+	}
+
+	return nil
+}
+
+// GetSelected returns the terminal command selected during the parse phase, or
+// nil if the command hasn't been parsed.
+func (cmd *Command) GetSelected() *Command {
+	if cmd.selected == nil {
+		return nil
+	}
+
+	if cmd.selected == cmd {
+		return cmd
+	}
+
+	return cmd.selected.GetSelected()
+}
+
+// GetParent returns the parent command of this command, or nil if a parent
+// hasn't been set. Parents are set during the parse phase, but only for
+// commands which are traversed.
+func (cmd *Command) GetParent() *Command {
+	return cmd.parent
+}
```

Comment:
The intent here is to help usage functions that want to include [information from preceding commands in the tree](https://github.com/peterbourgon/ff/issues/72).